Add Documentation to the mfe_deployer role (#6171)
edx/configuration: playbooks/roles/mfe_deployer/README.rst

        - **npm_overrides**: To indicate the list of npm overrides that will be installed. Used for branding, See the `Developer Documentation`_ for more details. By default it takes the value of *MFE_DEPLOY_NPM_OVERRIDES*.
        - **env_extra**: To define custom environment variables that will be used to build this MFE. By default it takes the value of *MFE_DEPLOY_ENVIRONMENT_EXTRA*.
        - **public_path**: To define the path where the MFE will be deployed. This requires tht the MFEs have compatibility with deployments in subdirectories. Regarding to the openedx ecosystem this means that the MFE should be using frontend-build>=1.3.2 and frontend-platform>=1.6.1. By default it takes the value of *MFE_DEPLOY_ENVIRONMENT_EXTRA*.
        - **site_name**: Used to define the Environment SITE_NAME, used to build the MFE. By default it takes the value of *MFE_DEPLOY_SITE_NAME*.
        - **standalone_nginx**: To indicate if the MFE will be deployed in a separated nginx file or if it will be in a shared nginx file with the other MFEs, by default it takes the value of *MFE_DEPLOY_STANDALONE_NGINX*.

Deployment using subdirectories
_______________________________

> By default, the mfe_deployer role is configured to deploy the microfrontends using subdirectories, the most important configurations to enable that are **MFES** and **MFE_DEPLOY_STANDALONE_NGINX**.  The *MFES* variable by default is defined as follows:

Reviewer: The [wording below](https://github.com/edx/configuration/pull/6171/files#diff-0539a91cf83ea97c0637aecf583609aae0a484c76158abc8809b4909d424df1dR56) is confusing and duplicates some information contained here.  Could these be simplified to something like this:

```suggestion
By default `MFE_DEPLOY_STANDALONE_NGINX` is false, which means that all the microfrontends defined in the `MFES` configuration will be deployed to different `public_path`s under the same domain (specifically, `MFE_DEPLOY_COMMON_HOSTNAME`).
```

>     MFES:

Reviewer: I don't think it's a good idea to repeat the default configuration here. Instead, better to show complete example config which deploys the default MFEs to subdirectories of a single subdomain and hooks them up to the LMS, e.g. 

```yaml
MFE_BASE: "mfe.{{ EDXAPP_LMS_BASE }}"

EDXAPP_SESSION_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
EDXAPP_CSRF_COOKIE_SECURE: true
EDXAPP_SESSION_COOKIE_SECURE: true
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_NAME: "cross-domain-cookie-mfe"

# Whitelisting
EDXAPP_CORS_ORIGIN_WHITELIST:
- "{{ EDXAPP_CMS_BASE }}"
- "{{ MFE_BASE }}"

EDXAPP_CSRF_TRUSTED_ORIGINS:
- "{{ MFE_BASE }}"

EDXAPP_LOGIN_REDIRECT_WHITELIST:
- "{{ EDXAPP_CMS_BASE }}"
- "{{ MFE_BASE }}"

# MFE Links
EDXAPP_LMS_WRITABLE_GRADEBOOK_URL: https://{{ MFE_BASE}}/gradebook
EDXAPP_PROFILE_MICROFRONTEND_URL: https://{{ MFE_BASE}}/profile/u/
EDXAPP_ACCOUNT_MICROFRONTEND_URL: https://{{ MFE_BASE}}/account
```

        public_path: "/account/"

On the other hand, **MFE_DEPLOY_STANDALONE_NGINX** is false by default in the mfe_deployer role which means that all the microfrontends defined in the **MFES** configuration will be deployed under the same domain (defined in the **MFE_DEPLOY_COMMON_HOSTNAME** configuration). If **MFE_DEPLOY_STANDALONE_NGINX** is false, please make sure that public_path starts and ends with a '/' and also check that the public_path configuration is not repeated for different microfrontends.

If we are deploying using subdirectories, it is necessary to set the **MFE_BASE** with the shared domain for the microfrontends.

Deployment using subdomains
___________________________

If we want to deploy the microfrontends in different subdomains, we should turn on the **MFE_DEPLOY_STANDALONE_NGINX**. As an example, this configurations are enough to deploy the profile, gradebook and account microfrontends in a different subdomain.

    MFES:
      - name: profile
        repo: frontend-app-profile
        public_path: "/"
      - name: gradebook
        repo: frontend-app-gradebook
        public_path: "/"
      - name: account
        repo: frontend-app-account

>         public_path: "/"

Reviewer: This code block and the ones above are still in MD format, see [rendered file](https://github.com/edx/configuration/blob/d78deab850f712bdf3a8dd720dd0b8a87e2062a0/playbooks/roles/mfe_deployer/README.rst#deployment-using-subdirectories).

>     MFE_DEPLOY_STANDALONE_NGINX: true

Reviewer: I think it's clearer to include the whitelisting and CSRF variables required for this example here, instead of just discussing them below.

```yaml
EDXAPP_SESSION_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
EDXAPP_CSRF_COOKIE_SECURE: true
EDXAPP_SESSION_COOKIE_SECURE: true
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_NAME: "cross-domain-cookie-mfe"

# Whitelisting
EDXAPP_LOGIN_REDIRECT_WHITELIST:
- "{{ EDXAPP_CMS_BASE }}"
- "profile.{{ EDXAPP_LMS_BASE }}"
- "gradebook.{{ EDXAPP_LMS_BASE }}"
- "account.{{ EDXAPP_LMS_BASE }}"

EDXAPP_CORS_ORIGIN_WHITELIST:
- "{{ EDXAPP_CMS_BASE }}"
- "profile.{{ EDXAPP_LMS_BASE }}"
- "gradebook.{{ EDXAPP_LMS_BASE }}"
- "account.{{ EDXAPP_LMS_BASE }}"

EDXAPP_CSRF_TRUSTED_ORIGINS:
- "{{ EDXAPP_CMS_BASE }}"
- "profile.{{ EDXAPP_LMS_BASE }}"
- "gradebook.{{ EDXAPP_LMS_BASE }}"
- "account.{{ EDXAPP_LMS_BASE }}"

# MFE Links
EDXAPP_LMS_WRITABLE_GRADEBOOK_URL: "https://gradebook.{{ EDXAPP_LMS_BASE}}/"
EDXAPP_PROFILE_MICROFRONTEND_URL: "https://profile.{{ EDXAPP_LMS_BASE}}/"
EDXAPP_ACCOUNT_MICROFRONTEND_URL: "https://account.{{ EDXAPP_LMS_BASE}}/"
```

_____________________

As described previously, the compilation of the MFEs is done in the *mfe* role, so some configurations cannot be overridden from *mfe_deployer*. You can see the list of all the default environment configuration in the defaults of the *mfe* role. The following variables can be overridden for all MFEs, but not individually: *MFE_MARKETING_SITE_BASE_URL*, *MFE_ENTERPRISE_MARKETING_UTM_SOURCE*, *MFE_ENTERPRISE_MARKETING_UTM_CAMPAIGN*, and *MFE_ENTERPRISE_MARKETING_FOOTER_UTM_MEDIUM* for configuration related to the marketing site and *MFE_NEW_RELIC_APP_ID* and *MFE_NEW_RELIC_LICENSE_KEY* in order to configure the newrelic integration. The default environment variables are defined in the `MFE_ENVIRONMENT_DEFAULT`_ configuration.

LMS Configuration
_________________

The deployment of the MFEs to the appservers will not be enough to have them working properly. Most of them require communication with the LMS, so it is necessary to configure the LMS to accept communication from the MFEs.

The principal configurations that are needed in ansible are: *EDXAPP_CORS_ORIGIN_WHITELIST*, *EDXAPP_CSRF_TRUSTED_ORIGINS*, *EDXAPP_LOGIN_REDIRECT_WHITELIST*. 

> They should contain the domain of the MFEs so that the LMS accepts their requests.

Reviewer: Worth mentioning the MFE URL variables that you also need to set so the LMS uses the default `MFES`: 
`EDXAPP_LMS_WRITABLE_GRADEBOOK_URL`, `EDXAPP_PROFILE_MICROFRONTEND_URL`,
`EDXAPP_ACCOUNT_MICROFRONTEND_URL`.

Comprehensive Example of a deployment using subdirectories
__________________________________________________________

.. code-block:: yaml

  MFE_BASE: "mfe.{{ EDXAPP_LMS_BASE }}"

  MFES:
    - name: profile
      repo: frontend-app-profile
      public_path: "/profile/"
    - name: gradebook
      repo: frontend-app-gradebook
      public_path: "/gradebook/"
    - name: account
      repo: frontend-app-account
      public_path: "/account/"

  MFE_DEPLOY_STANDALONE_NGINX: false
  MFE_DEPLOY_COMMON_HOSTNAME: '{{ MFE_BASE }}'

  EDXAPP_SESSION_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
  EDXAPP_CSRF_COOKIE_SECURE: true
  EDXAPP_SESSION_COOKIE_SECURE: true
  EDXAPP_ENABLE_CORS_HEADERS: true
  EDXAPP_ENABLE_CROSS_DOMAIN_CSRF_COOKIE: true
  EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_DOMAIN: ".{{ EDXAPP_LMS_BASE }}"
  EDXAPP_CROSS_DOMAIN_CSRF_COOKIE_NAME: "cross-domain-cookie-mfe"

  EDXAPP_CORS_ORIGIN_WHITELIST:
    - "{{ EDXAPP_CMS_BASE }}"
    - "{{ MFE_BASE }}"

  EDXAPP_CSRF_TRUSTED_ORIGINS:
    - "{{ MFE_BASE }}"

  EDXAPP_LOGIN_REDIRECT_WHITELIST:
    - "{{ EDXAPP_CMS_BASE }}"
    - "{{ MFE_BASE }}"

> 

Reviewer: Only thing missing for `CORS` is:

```yaml
EDXAPP_FEATURES:
  ENABLE_CORS_HEADERS: true
```

Author: I think that unless you are overriding the default EDXAPP_FEATURES configuration, it should be being working with `EDXAPP_ENABLE_CORS_HEADERS`

https://github.com/edx/configuration/blob/a2c4e00ec157b762ea5786111db22bf34e005742/playbooks/roles/edxapp/defaults/main.yml#L327

Reviewer: Cool, yep, that's better!

  # MFE Links

>   EDXAPP_LMS_WRITABLE_GRADEBOOK_URL: 'https://{{ MFE_BASE}}/gradebook'

Reviewer: FYI unfortunately, there's a bug in edx-platform that prevents the gradebook link from working when deployed with subdirectories.

[The code](https://github.com/edx/edx-platform/blob/151bd136667d27bf85256e382ed5b5ea0144e00f/lms/djangoapps/instructor/views/instructor_dashboard.py#L574) uses [`urljoin`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin), which uses only the hostname from `WRITABLE_GRADEBOOK_URL`, and removes the `/gradebook` path.

So the above example produces e.g., `https://mfe.my.lms.com/course-v1:edX+DemoX+Demo_Course` instead of `https://mfe.my.lms.com/gradebook/course-v1:edX+DemoX+Demo_Course`.

Not something that needs to be fixed here obviously, but noting it as something we'll need to fix to use subdirectories with this MFE.

        - **npm_overrides**: To indicate the list of npm overrides that will be installed. Used for branding, See the `Developer Documentation`_ for more details. By default it takes the value of ``MFE_DEPLOY_NPM_OVERRIDES``.
        - **env_extra**: To define custom environment variables that will be used to build this MFE. By default it takes the value of ``MFE_DEPLOY_ENVIRONMENT_EXTRA``.
        - **public_path**: To define the path where the MFE will be deployed. This requires that the MFEs have compatibility with deployments in subdirectories. Regarding to the openedx ecosystem this means that the MFE should be using frontend-build>=1.3.2 and frontend-platform>=1.6.1. By default it takes the value of ``MFE_DEPLOY_PUBLIC_PATH``.
        - **site_name**: Used to define the Environment SITE_NAME, used to build the MFE. By default it takes the value of ``MFE_DEPLOY_SITE_NAME``.
        - **standalone_nginx**: To indicate if the MFE will be deployed in a separated nginx file or if it will be in a shared nginx file with the other MFEs, by default it takes the value of ``MFE_DEPLOY_STANDALONE_NGINX``.

Deployment using subdirectories
_______________________________

> By default ``MFE_DEPLOY_STANDALONE_NGINX`` is false, which means that all the microfrontends defined in the ``MFES`` configuration will be deployed to different ``public_path``'s under the same domain (specifically, ``MFE_DEPLOY_COMMON_HOSTNAME``). On the other hand, ``MFE_DEPLOY_STANDALONE_NGINX`` is false by default in the mfe_deployer role which means that all the microfrontends defined in the ``MFES`` configuration will be deployed under the same domain (defined in the ``MFE_DEPLOY_COMMON_HOSTNAME`` configuration).

Reviewer: nit: the same thing is said twice here, so you can remove the second sentence, starting with "On the other hand...".

LMS Configuration
_________________

The deployment of the MFEs to the appservers will not be enough to have them working properly. Most of them require communication with the LMS, so it is necessary to configure the LMS to accept communication from the MFEs.

The principal configurations that are needed in ansible are: ``EDXAPP_CORS_ORIGIN_WHITELIST``, ``EDXAPP_CSRF_TRUSTED_ORIGINS``, ``EDXAPP_LOGIN_REDIRECT_WHITELIST``. 
They should contain the domain of the MFEs so that the LMS accepts their requests.

It is also necessary to enable the features ENABLE_CORS_HEADERS and ENABLE_CROSS_DOMAIN_CSRF_COOKIE. They can be enabled in Koa with ``EDXAPP_ENABLE_CORS_HEADERS``, ``EDXAPP_ENABLE_CROSS_DOMAIN_CSRF_COOKIE``.

It is also necessary to have configured JWT properly in the LMS. You can use the generate_jwt_signing_key command to generate the signing key. See `decision record about asymmetric JWT`_ for more details.

> For each MFE, it might be certain configurations that also need to be changed according to the URLs of the MFE, for instance, for the gradebook, progile and account MFE we need to set ``EDXAPP_LMS_WRITABLE_GRADEBOOK_URL``, ``EDXAPP_PROFILE_MICROFRONTEND_URL`` and ``EDXAPP_ACCOUNT_MICROFRONTEND_URL`` with their respective URLs.

Reviewer: nit: "progile" is misspelling of "profile"

mfe_deployer
############

Overview
--------

The purpose of this document is to provide an overview of how micro-frontend applications (MFEs) can be deployed using the ansible roles from the edx/configuration repository.

For a generic overview of the specific steps that are required for the deployment, please see the [Developer Documentation](https://edx.readthedocs.io/projects/edx-developer-docs/en/latest/developers_guide/micro_frontends_in_open_edx.html).

> In order to deploy MFEs with ansible, there were created by the community two roles `mfe` and `mfe_deployer`:

Reviewer 2: ```suggestion
Two ansible roles were created by the community to deploy MFEs using `edx/configuration`:
```

> - `mfe`: The `mfe` role is the base role for the deployments, in that role a specific mfe is cloned, build and then its files are served using nginx. This role internally, follows the steps for building and deployiment of MFEs described in the developer documentation, with the exception that it does not uses tubular. Ansible users should not use directly this role, and instead use the `mfe_deployer` role.

Reviewer 2: ```suggestion
- `mfe`: The `mfe` role is the base role for the deployments where a single mfe is cloned, built and set up to be served using nginx from the appserver. This role internally, follows the steps for building and deployment of MFEs described in the developer documentation, with the exception that it does not uses tubular. Ansible users should not use directly this role, and instead use the `mfe_deployer` role.
```

- `mfe_deployer`: The `mfe_deployer` role is used to deploy several MFEs in a programmatic way. Internally, this role, make a call to the base `mfe` role for each MFE that is intended to be deployed.

Configuration & Deployment

> --------------------------

Reviewer 2: I think it's worth mentioning here that MFEs can be deployed in one of two ways:

1. Using the same subdomain, but served from different directories.
2. Using separate subdomains.

And mention the variables that control which deployment type is configured.
We can also mention what the default values deploy, and a example of variables to deploy on separate subdomains vs on the same domain.

> When running this role, you'' need to set:

Reviewer 2: ```suggestion
When running this role, you'll need to set the following variables:
```

> - `MFES`: list that contains each one of the MFEs that will be deployed. Each MFE should contain a `name` and a `repo`. Optionally, it can be defined for each MFE:

Reviewer 2: ```suggestion
- `MFES`: list that contains each one of the MFEs that will be deployed. Each MFE should contain a `name` and a `repo`. Optionally, the extra parameters below can be defined for each MFE:
```

        - `env_extra`: To define custom environment variables that will be used to build this MFE. By default it takes the value of `MFE_DEPLOY_ENVIRONMENT_EXTRA`.
        - `public_path`: To define the path where the MFE will be deployed. This requires tht the MFEs have compatibility with deployments in subdirectories. Regarding to the openedx ecosystem this means that the MFE should be using frontend-build>=1.3.2 and frontend-platform>=1.6.1. By default it takes the value of `MFE_DEPLOY_ENVIRONMENT_EXTRA`.
        - `site_name`: Used to define the Environment SITE_NAME, used to build the MFE. By default it takes the value of `MFE_DEPLOY_SITE_NAME`.
        - `standalone_nginx`: To indicate if the MFE will be deployed in a separated nginx file or if it will be in a shared nginx file with the other MFEs, by default it takes the value of `MFE_DEPLOY_STANDALONE_NGINX`.

> By default, MFES is defined as:

Reviewer 2: ```suggestion
By default, MFES variable is defined as follows:
```

```yaml

MFES:
  - name: profile
    repo: frontend-app-profile
    public_path: "/profile/"
  - name: gradebook
    repo: frontend-app-gradebook
    public_path: "/gradebook/"
  - name: account
    repo: frontend-app-account
    public_path: "/account/
```

> So it will deploy three MFEs each one in a different subdirectory. Since the default value of `MFE_DEPLOY_STANDALONE_NGINX` is false. The three MFE will be deployed in the same nginx file.

Reviewer 2: ```suggestion
This configuration will deploy three MFEs, each in a different subdirectory. And since the default value of `MFE_DEPLOY_STANDALONE_NGINX` is false,  the three MFEs will be deployed in the same nginx file.
```

Reviewer 2: @morenol I'm not sure what you mean by `deployed in the same Nginx file` here can you elaborate a bit more?

Also, the default value of `MFE_DEPLOY_STANDALONE_NGINX` is true, see https://github.com/edx/configuration/blob/b828714637f55fc2c11faec8e59b287c7fc730fd/playbooks/roles/mfe/defaults/main.yml#L136.

Author: You are right, I have to add a better description, I am trying to say that the nginx configuration for all  of the MFEs is placed in the same in 

/etc/nginx/sites_enabled/

I agree that I have to explain that in a more descriptive way. On the other hand it is true in that role, but in the mfe_deployer (which is the role that should be used by the edx/configuration  users) it is false.

> As it was described previously, the compilation of the MFEs is done in the `mfe` role. Some configurations cannot be override as of today from the mfe_deployer, you can see the list of all the default environment configuration in the defaults of the mfe role. For instance, it can be defined `MFE_MARKETING_SITE_BASE_URL`, `MFE_ENTERPRISE_MARKETING_UTM_SOURCE`, `MFE_ENTERPRISE_MARKETING_UTM_CAMPAIGN`, and `MFE_ENTERPRISE_MARKETING_FOOTER_UTM_MEDIUM` for configuration related to the marketing site and `MFE_NEW_RELIC_APP_ID` and `MFE_NEW_RELIC_LICENSE_KEY` in order to configure the newrelic integration.

Reviewer 2: ```suggestion
As described previously, the compilation of the MFEs is done in the `mfe` role, so some configurations cannot be overridden from `mfe_deployer`. You can see the list of all the default environment configuration in the defaults of the `mfe` role. The following variables can be overridden for all MFEs, but not individually: `MFE_MARKETING_SITE_BASE_URL`, `MFE_ENTERPRISE_MARKETING_UTM_SOURCE`, `MFE_ENTERPRISE_MARKETING_UTM_CAMPAIGN`, and `MFE_ENTERPRISE_MARKETING_FOOTER_UTM_MEDIUM` for configuration related to the marketing site and `MFE_NEW_RELIC_APP_ID` and `MFE_NEW_RELIC_LICENSE_KEY` in order to configure the newrelic integration.
```

> As it was described previously, the compilation of the MFEs is done in the `mfe` role. Some configurations cannot be override as of today from the mfe_deployer, you can see the list of all the default environment configuration in the defaults of the mfe role. For instance, it can be defined `MFE_MARKETING_SITE_BASE_URL`, `MFE_ENTERPRISE_MARKETING_UTM_SOURCE`, `MFE_ENTERPRISE_MARKETING_UTM_CAMPAIGN`, and `MFE_ENTERPRISE_MARKETING_FOOTER_UTM_MEDIUM` for configuration related to the marketing site and `MFE_NEW_RELIC_APP_ID` and `MFE_NEW_RELIC_LICENSE_KEY` in order to configure the newrelic integration.

Reviewer 2: @morenol Can you add a link to the MFE default variable here?

LMS Configuration
-----------------

> The deployment of the MFEs by themselves will not be enought to have them working properly. Most of them require communication with the LMS, so it is necessary to configure the LMS to accept communication from the MFEs.

Reviewer 2: ```suggestion
The deployment of the MFEs to the appservers will not be enough to have them working properly. Most of them require communication with the LMS, so it is necessary to configure the LMS to accept communication from the MFEs.
```

> The principal configurations that are needed in ansible are: `EDXAPP_CORS_ORIGIN_WHITELIST, EDXAPP_CSRF_TRUSTED_ORIGINS, EDXAPP_LOGIN_REDIRECT_WHITELIST`. They should contain the domain of the MFEs in order to configure the LMS to accepts requests from the MFE.

Reviewer 2: ```suggestion
The principal configurations that are needed in ansible are: `EDXAPP_CORS_ORIGIN_WHITELIST, EDXAPP_CSRF_TRUSTED_ORIGINS, EDXAPP_LOGIN_REDIRECT_WHITELIST`. 
They should contain the domain of the MFEs so that the LMS accepts their requests.
```

It is also necessary to enable the features ENABLE_CORS_HEADERS and ENABLE_CROSS_DOMAIN_CSRF_COOKIE. They can be enabled in Koa with `EDXAPP_ENABLE_CORS_HEADERS` `EDXAPP_ENABLE_CROSS_DOMAIN_CSRF_COOKIE`.

> It is also necessary to have configured JWT properly in the LMS.

Reviewer 2: By this you mean that the JWK Key pairs are configured correctly? Is there any specific configuration?

Author: Yes, I think that it is not needed a specific configuration for them, only the configurations that we can generate through this command https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/oauth_dispatch/management/commands/generate_jwt_signing_key.py/

Other than that, I need to check better which are the required configurations for the CORS configurations.
